Sources/CodeEditTextView/TreeSitter/TreeSitterClient.swift

```diff
@@ -75,143 +114,164 @@ final class TreeSitterClient: HighlightProviding {
+            layers.remove(at: layerIdx)
         }
+
         completion(rangeSet)
     }
 
+    /// Initiates a highlight query.
+    /// - Parameters:
+    ///   - textView: The text view to use.
+    ///   - range: The range to limit the highlights to.
+    ///   - completion: Called when the query completes.
     func queryHighlightsFor(textView: HighlighterTextView,
                             range: NSRange,
-                            completion: @escaping (([HighlightRange]) -> Void)) {
-        // Make sure we dont accidentally change the tree while we copy it.
-        self.semaphore.wait()
-        guard let tree = self.tree?.copy() else {
-            // In this case, we don't have a tree to work with already, so we need to make it and try to
-            // return some highlights
-            createTree(textView: textView)
-
-            // This is slightly redundant but we're only doing one check.
-            guard let treeRetry = self.tree?.copy() else {
-                // Now we can return nothing for real.
-                self.semaphore.signal()
-                completion([])
-                return
+                            completion: @escaping ((([HighlightRange]) -> Void))) {
+        var highlights: [HighlightRange] = []
+        var injectedSet = IndexSet(integersIn: range)
+
+        for layer in layers where layer.id != primaryLayer {
+            // Query injected only if a layer's ranges intersects with `range`
+            for layerRange in layer.ranges {
+                if let rangeIntersection = range.intersection(layerRange) {
+                    highlights.append(
+                        contentsOf: queryLayerHighlights(layer: layer,
+                                                         textView: textView,
+                                                         range: rangeIntersection)
+                    )
+
+                    injectedSet.remove(integersIn: rangeIntersection)
+                }
             }
-            self.semaphore.signal()
-
-            _queryColorsFor(tree: treeRetry, range: range, completion: completion)
-            return
         }
 
-        self.semaphore.signal()
-
-        _queryColorsFor(tree: tree, range: range, completion: completion)
-    }
-
-    private func _queryColorsFor(tree: Tree,
-                                 range: NSRange,
-                                 completion: @escaping (([HighlightRange]) -> Void)) {
-        guard let rootNode = tree.rootNode else {
-            completion([])
-            return
+        // Query primary for any ranges that weren't used in the injected layers.
+        for range in injectedSet.rangeView {
+            highlights.append(contentsOf: queryLayerHighlights(layer: layers[0],
+                                                               textView: textView,
+                                                               range: NSRange(range)))
```

Comment:
### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
            highlights.append(contentsOf: queryLayerHighlights(
                layer: layers[0],
                textView: textView,
                range: NSRange(range)
            ))
```